### PR TITLE
Cosmology read & write method for AAS MRT

### LINKED ISSUE #14871: Cosmology read & write method for AAS MRT
The AAS MRT (https://docs.astropy.org/en/stable/io/unified.html#built-in-table-readers-writers) format would be a nice addition to cosmology I/O.

_Originally posted by @nstarman in https://github.com/astropy/astropy/issues/14837#issuecomment-1563429280_

### PR DESCRIPTION
**Pull request by Mubin17**
### Description

This pull request is to address #14871

Fixes #14871

Registered the read_mrt() and write_mrt() methods with the Cosmology class, enabling the import and export of cosmological data to and from MRT files.

## Writing Example:
```python
from astropy.cosmology import Planck18
Planck18.write("example_cosmology.mrt", overwrite=True, format="mrt")
```
Output:
```plaintext
Title:
Authors:
Table:
================================================================================
Byte-by-byte Description of file: table.dat
--------------------------------------------------------------------------------
 Bytes Format Units  Label     Explanations
--------------------------------------------------------------------------------
 1-13  A13          ---    cosmology Description of cosmology
15-22  A8           ---    name      Description of name
24-28  F5.2   km.Mpc-1.s-1 H0        [67.66/67.66] Hubble constant as an
                              `~astropy.units.Quantity` at z=0.
30-36  F7.5         ---    Om0       [0.3/0.31] Omega matter; matter
                              density/critical density at z=0.
38-43  F6.4         K      Tcmb0     [2.72/2.73] Temperature of the CMB as
                              `~astropy.units.Quantity` at z=0.
45-49  F5.3         ---    Neff      [3.04/3.05] Number of effective neutrino
                              species.
51-53  F3.1         eV     mnu0      Description of mnu0
55-57  F3.1         eV     mnu1      Description of mnu1
59-62  F4.2         eV     mnu2      [0.06/0.06] Description of mnu2
64-70  F7.5         ---    Ob0       [0.04/0.05] Omega baryon; baryonic matter
                              density/critical density at z=0.
--------------------------------------------------------------------------------
Notes:
--------------------------------------------------------------------------------
FlatLambdaCDM Planck18 67.66 0.30966 2.7255 3.046 0.0 0.0 0.06 0.04897
```
## Reading Example

```python
from astropy.cosmology import Cosmology
Cosmology.read('example_cosmology.mrt', format='mrt')
```
Output:
```plaintext
FlatLambdaCDM(name="Planck18", H0=67.66 km / (Mpc s), Om0=0.30966, Tcmb0=2.7255 K, Neff=3.046, m_nu=[0.   0.   0.06] eV, Ob0=0.04897)
```

### CODE REVIEW

File: astropy/cosmology/io/mrt.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+from astropy.cosmology.connect import readwrite_registry
+from astropy.cosmology.core import Cosmology
+from astropy.table import QTable
+from astropy.table.serialize import represent_mixins_as_columns
+
+from .table import from_table, to_table
+
+
+def read_mrt(filename, index=None, *, move_to_meta=False, cosmology=None, **kwargs):
+    """Read a |Cosmology| from an MRT file.
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:10]:
```suggestion
    """Read a |Cosmology| from a MRT file.
```

File: astropy/cosmology/io/mrt.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    Parameters
+    ----------
+    filename : path-like or file-like
+        The file from which to read the MRT data.
+    index : int or str or None, optional
+        Needed to select the row in tables with multiple rows. ``index`` can be
+        an integer for the row number or, if the table is indexed by a column,
+        the value of that column. If the table is not indexed and ``index`` is a
+        string, the "name" column is used as the indexing column.
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:20]: I don't think MRT tables preserve which column is an indexing column.

```suggestion
    index : int or str or None, optional
        Needed to select the row in tables with multiple rows. ``index`` can be
        an integer for the row number or, if ``index`` is a
        string, the "name" column is used as the indexing column.
```

File: astropy/cosmology/io/mrt.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        The 'format' is set to 'mrt', regardless of input.
+
+    Returns
+    -------
+    |Cosmology| subclass instance
+
+    Raises
+    ------
+    ValueError
+        If the keyword argument 'format' is given and is not "mrt".
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:44]:
```suggestion
        If the keyword argument 'format' is given and is not "ascii.mrt".
```

File: astropy/cosmology/io/mrt.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    # Check that the format is 'mrt' (or not specified)
+    format = kwargs.pop("format", "mrt")
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:47]:
```suggestion
    format = kwargs.pop("format", "ascii.mrt")
```

File: astropy/cosmology/io/mrt.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    if format != "mrt":
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:48]:
```suggestion
    if format != "ascii.mrt":
```

File: astropy/cosmology/io/mrt.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        raise ValueError(f"format must be 'mrt', not {format}")
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:49]:
```suggestion
        raise ValueError(f"format must be 'ascii.mrt', not {format}")
```

File: astropy/cosmology/io/mrt.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+    # Reading is handled by `QTable`.
+    table = QTable.read(filename, format="mrt", **kwargs)
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:52]:
```suggestion
    table = QTable.read(filename, format="ascii.mrt", **kwargs)
```

File: astropy/cosmology/io/mrt.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+    # Create a single column named 'm_nu' by combining the values from the 'mnu0', 'mnu1', and 'mnu2' columns
+    m_nu_data = []
+    for i in ("mnu0", "mnu1", "mnu2"):
+        column_data = table[i][0]
+        m_nu_data.append(column_data.value)
+        table.remove_column(i)
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:59]: With `represent_mixins_as_columns(table_main)`, does this serialize without breaking apart the neutrino mass?

File: astropy/cosmology/io/mrt.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    table_main = to_table(cosmology, cls=cls, cosmology_in_meta=False)
+    table = represent_mixins_as_columns(table_main)
+
+    # Replace the m_nu column with three columns with names 'mnu0', 'mnu1', 'mnu2'
+    m_nu = table_main["m_nu"]
+    table.remove_column("m_nu")
+    for i in range(len(m_nu[0])):
+        column_name = f"mnu{i}"
+        column_data = m_nu[0][i]
+        table.add_column(column_data, name=column_name, index=-2)
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:107]: With `represent_mixins_as_columns(table_main)`, does this serialize without breaking apart the neutrino mass?

File: astropy/cosmology/io/mrt.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        The |Table| class to use when writing the MRT file. Default is |QTable|.
+    **kwargs : Any
+        Additional keyword arguments passed to ``cls.write``.
+
+    Raises
+    ------
+    TypeError
+        If kwarg (optional) 'cls' is not a subclass of `astropy.table.Table`
+    ValueError
+        If the keyword argument 'format' is given and is not "mrt".
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:90]:
```suggestion
        If the keyword argument 'format' is given and is not "ascii.mrt".
```

File: astropy/cosmology/io/mrt.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    # Check that the format is 'mrt' (or not specified)
+    format = kwargs.pop("format", "mrt")
+    if format != "mrt":
+        raise ValueError(f"format must be 'mrt', not {format}")
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:95]:
```suggestion
    format = kwargs.pop("format", "ascii.mrt")
    if format != "ascii.mrt":
        raise ValueError(f"format must be 'ascii.mrt', not {format}")
```

File: astropy/cosmology/io/mrt.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+    # Replace the m_nu column with three columns with names 'mnu0', 'mnu1', 'mnu2'
+    m_nu = table_main["m_nu"]
+    table.remove_column("m_nu")
+    for i in range(len(m_nu[0])):
+        column_name = f"mnu{i}"
+        column_data = m_nu[0][i]
+        table.add_column(column_data, name=column_name, index=-2)
+
+    table.write(file, overwrite=overwrite, format="mrt", **kwargs)
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:109]:
```suggestion
    table.write(file, overwrite=overwrite, format="ascii.mrt", **kwargs)
```

File: astropy/cosmology/io/mrt.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+
+# ===================================================================
+# Register
+
+readwrite_registry.register_reader("mrt", Cosmology, read_mrt)
+readwrite_registry.register_writer("mrt", Cosmology, write_mrt)
```
Comment (nstarman) [inline astropy/cosmology/io/mrt.py:116]:
```suggestion
readwrite_registry.register_reader("ascii.mrt", Cosmology, read_mrt)
readwrite_registry.register_writer("ascii.mrt", Cosmology, write_mrt)
```

**Review by nstarman — CHANGES_REQUESTED** [review]
Thanks @Mubin17 for the PR. Most things look good. I have just 2 corrections"
1. the format should be "ascii.mrt", in line with how it's registered with Table
2. with `represent_mixins_as_columns` do we still need to make separate neutrino columns?

File: astropy/cosmology/io/mrt.py (hunk 9 above)
Comment (Mubin17) [reply astropy/cosmology/io/mrt.py:107]: I may be completely wrong in my understanding of how serializing is supposed to work in this context. But I believe ```represent_mixins_as_columns(table_main)```, does not serialize the table without breaking apart the neutrino mass, and here is why:

```python
from astropy.cosmology import Planck18
from astropy.table.serialize import represent_mixins_as_columns
table_main = Planck18.to_format("astropy.table")
table = represent_mixins_as_columns(table_main)

m_nu_table_main = table_main['m_nu'] # Get the m_nu column from table_main
m_nu_table = table['m_nu']  # Get the m_nu column from table

N = len(m_nu_table[0])  # Get the number of elements in m_nu

for i in range(N):
    column_data_table = m_nu_table[0][i]
    print(column_data_table, type(column_data_table))
    
for i in range(N):
    column_data_table_main = m_nu_table_main[0][i]
    print(column_data_table_main, type(column_data_table_main))
```
Output:
```plaintext
0.0 <class 'numpy.float64'>
0.0 <class 'numpy.float64'>
0.06 <class 'numpy.float64'>
0.0 eV <class 'astropy.units.quantity.Quantity'>
0.0 eV <class 'astropy.units.quantity.Quantity'>
0.06 eV <class 'astropy.units.quantity.Quantity'>
```
The way I see it is that, for serialization to work correctly the outputs must be the same.

**Comment by nstarman** [comment]
@Mubin17, the tests are failing. It looks like many are related to this PR. If you need help, feel free to ping me for questions.

File: astropy/cosmology/io/mrt.py (hunk 9 above)
Comment (nstarman) [reply astropy/cosmology/io/mrt.py]: You're right that `represent_mixins_as_columns` doesn't help here. I'll open an issue to see if we can extend `represent_mixins_as_columns` to make structured arrays into a series of columns.

Comment (nstarman) [reply astropy/cosmology/io/mrt.py]: ping @taldcroft

Comment (nstarman) [reply astropy/cosmology/io/mrt.py]: I opened https://github.com/astropy/astropy/issues/14913 for this change.

**Comment by Mubin17** [comment]
@nstarman do have any insight on what is triggering this error, ```m_nu[0]``` is iterable and it is not a scalar so I don't understand why the error?

```shell
self = <astropy.cosmology.tests.test_core.TestCosmology object at 0x1186f38b0>, read = <astropy.cosmology.connect.CosmologyRead object at 0x118e04c40>
write = <astropy.cosmology.connect.CosmologyWrite object at 0x118e06140>
tmp_path = PosixPath('/private/var/folders/t3/4lbhlq652z16wh1ldqsfr04m0000gn/T/pytest-of-mubinmanasia/pytest-8/test_to_mrt_bad_index0')

    def test_to_mrt_bad_index(self, read, write, tmp_path):
        """Test if argument ``index`` is incorrect"""
        fp = tmp_path / "test_to_mrt_bad_index.mrt"
    
>       write(fp, format="ascii.mrt")

cosmology/io/tests/test_mrt.py:25: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cosmology/connect.py:126: in __call__
    self.registry.write(self._instance, *args, **kwargs)
io/registry/core.py:383: in write
    return writer(data, *args, **kwargs)
cosmology/io/mrt.py:127: in write_mrt
    cols, names = tuple(zip(*((m, f"m_nu[{i}]") for i, m in enumerate(m_nu[0]))))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Quantity 0.6 eV>

    def __iter__(self):
        if self.isscalar:
>           raise TypeError(
                f"'{self.__class__.__name__}' object with a scalar value is not"
                " iterable"
            )
E           TypeError: 'Quantity' object with a scalar value is not iterable

units/quantity.py:1275: TypeError
```

**Comment by nstarman** [comment]
So this is my fault. If you look at `astropy.cosmology.tests.test_core.TestCosmology`

```
        self._cls_args = dict(
            H0=70 * (u.km / u.s / u.Mpc), Tcmb0=2.7 * u.K, m_nu=0.6 * u.eV
        )
```

m_nu can be a scalar, which doesn't work with the code I pushed. m_nu needs to be listified first.

**Comment by Mubin17** [comment]
@nstarman I've listified the [m_nu](https://github.com/astropy/astropy/pull/14883/files#diff-8cf956e8b929bbf24e70781d969b71d71529bb4a5f42eb27451f960fc35745fbR387) parameter in the ```astropy.cosmology.tests.test_core.TestCosmology``` class and this fixed some of the failing tests. I also updated the index values from (-2, -2, -2) to (-4, -3, -2) to make sure ```m_un[0]```, ```m_un[1]```, and ```m_un[2]``` are written in the correct order.

There is one last set of failing tests that needs to be fixed:
```shell
self = <astropy.cosmology.flrw.tests.test_wpwazpcdm.TestwpwaCDM object at 0x12b6f0ac0>, cosmo_cls = <class 'astropy.cosmology.flrw.wpwazpcdm.wpwaCDM'>
cosmo = wpwaCDM(name="ABCMeta", H0=70.0 km / (Mpc s), Om0=0.27, Ode0=0.73, wp=-0.9, wa=0.2, zp=0.5 redshift, Tcmb0=3.0 K, Neff=3.04, m_nu=[0. 0. 0.] eV, Ob0=0.03)
read = <astropy.cosmology.connect.CosmologyRead object at 0x1693be8c0>, write = <astropy.cosmology.connect.CosmologyWrite object at 0x1693bfa00>
tmp_path = PosixPath('/private/var/folders/t3/4lbhlq652z16wh1ldqsfr04m0000gn/T/pytest-of-mubinmanasia/pytest-28/test_readwrite_mrt_instance9')

    def test_readwrite_mrt_instance(
        self, cosmo_cls, cosmo, read, write, tmp_path
    ):
        """Test cosmology -> ascii.mrt -> cosmology."""
        fp = tmp_path / "test_readwrite_mrt_instance.mrt"
    
        # ------------
        # To Table
    
        write(fp, format="ascii.mrt")
    
        # some checks on the saved file
        tbl = Table.read(fp, format="ascii.mrt")
        assert tbl["name"] == cosmo.name
    
        # ------------
        # From Table
    
        tbl["mismatching"] = "will error"
        tbl.write(fp, format="ascii.mrt", overwrite=True)
    
        # tests are different if the last argument is a **kwarg
        if tuple(cosmo._init_signature.parameters.values())[-1].kind == 4:
            got = read(fp, format="ascii.mrt")
    
            assert got.__class__ is cosmo_cls
            assert got.name == cosmo.name
            assert "mismatching" not in got.meta
    
            return  # don't continue testing
    
        # read with mismatching parameters errors
        with pytest.raises(TypeError, match="there are unused parameters"):
            read(fp, format="ascii.mrt")
    
        # unless mismatched are moved to meta
>       got = read(fp, format="ascii.mrt", move_to_meta=True)

cosmology/io/tests/test_mrt.py:89: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cosmology/connect.py:86: in __call__
    cosmo = self.registry.read(self._cls, *args, **kwargs)
io/registry/core.py:218: in read
    data = reader(*args, **kwargs)
cosmology/io/mrt.py:77: in read_mrt
    return from_table(
cosmology/io/table.py:138: in from_table
    return from_row(row, move_to_meta=move_to_meta, cosmology=cosmology)
cosmology/io/row.py:80: in from_row
    return from_mapping(mapping, move_to_meta=move_to_meta, cosmology=cosmology)
cosmology/io/mapping.py:114: in from_mapping
    return cosmology(*ba.args, **ba.kwargs)
cosmology/flrw/wpwazpcdm.py:141: in __init__
    self.zp = zp
cosmology/parameter.py:138: in __set__
    value = self.validate(cosmology, copy.deepcopy(value))
cosmology/parameter.py:186: in validate
    return self.fvalidate(cosmology, self, value)
cosmology/parameter.py:348: in _validate_with_unit
    value = u.Quantity(value, param.unit)
units/quantity.py:445: in __new__
    value = value.to(unit)
units/quantity.py:935: in to
    value = self._to_value(unit, equivalencies)
units/quantity.py:888: in _to_value
    return self.unit.to(
units/core.py:1195: in to
    return self._get_converter(Unit(other), equivalencies)(value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = UnrecognizedUnit(redshift), other = Unit("redshift"), equivalencies = []

    def _get_converter(self, other, equivalencies=None):
        self._normalize_equivalencies(equivalencies)
>       raise ValueError(
            f"The unit {self.name!r} is unrecognized.  It can not be converted "
            "to other units."
        )
E       ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.

units/core.py:2015: ValueError
------------------------------------------------------------------------------------------- Captured stdout call -------------------------------------------------------------------------------------------
cosmology   name       H0      Om0  Ode0 ... Neff m_nu[0] m_nu[1] m_nu[2] Ob0 
                  km / (Mpc s)           ...         eV      eV      eV       
--------- ------- ------------ ---- ---- ... ---- ------- ------- ------- ----
  wpwaCDM ABCMeta         70.0 0.27 0.73 ... 3.04     0.0     0.0     0.0 0.03
```
The same before this error is triggered whenever the read_mrt() method is called in tests

```shell
FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestwpwaCDM::test_readwrite_mrt_instance - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestwpwaCDM::test_readwrite_mrt_subclass_partial_info - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestwpwaCDM::test_readwrite_complete_info[ascii.mrt-False-True] - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestwpwaCDM::test_readwrite_from_subclass_complete_info[ascii.mrt-False-True] - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestFlatwpwaCDM::test_readwrite_mrt_instance - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestFlatwpwaCDM::test_readwrite_mrt_subclass_partial_info - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestFlatwpwaCDM::test_readwrite_complete_info[ascii.mrt-False-True] - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestFlatwpwaCDM::test_readwrite_from_subclass_complete_info[ascii.mrt-False-True] - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
```

**Comment by nstarman** [comment]
> Thanks for the insight @nstarman I've listified the [m_nu](https://github.com/astropy/astropy/pull/14883/files#diff-8cf956e8b929bbf24e70781d969b71d71529bb4a5f42eb27451f960fc35745fbR387) parameter in the `astropy.cosmology.tests.test_core.TestCosmology` class.

Sorry of the confusion. ``m_nu`` shouldn't be listified in the test, it should be fixed in ``write_mrt``. Astropy currently supports scalar `m_nu`, which is being tested.

**Comment by Mubin17** [comment]
@nstarman Thanks for the clarification. I was able listify the ```m_nu``` parameter by making the following changes to the existing ```write_mrt``` method:
```python
    # Replace the m_nu column with three columns with names 'm_nu_[i]'
    if "m_nu" in table.colnames:
        m_nu = table_main["m_nu"]
        table.remove_column("m_nu")
        m_nu = np.atleast_1d(m_nu) # Ensure m_nu is an array
        if m_nu.shape == (1,):
            table.add_column(m_nu[0], name="m_nu[0]", index=-2) # Add a single column 
        else:
            # Create multiple columns with appropriate names
            cols, names = tuple(zip(*((m, f"m_nu[{i}]") for i, m in enumerate(m_nu[0]))))
            table.add_columns(cols, names=names, indexes=(-4, -3, -2))
``` 
I ran tests on my local machine and they passed successfully.
Question: Do you know if I am allowed to use numpy here?

**Comment by nstarman** [comment]
> I ran tests on my local machine and they passed successfully.

Great!

> Question: Do you know if I am allowed to use numpy here?

Always! Numpy is a runtime dependency for Astropy

**Comment by Mubin17** [comment]
> Thanks for the insight @nstarman I've listified the [m_nu](https://github.com/astropy/astropy/pull/14883/files#diff-8cf956e8b929bbf24e70781d969b71d71529bb4a5f42eb27451f960fc35745fbR387) parameter in the `astropy.cosmology.tests.test_core.TestCosmology` class and this seem to fix some of the failing tests. However, I’m not sure if this is what you meant by the phrase "m_nu needs to be listified first". I have also updated the index values from (-2, -2, -2) to (-4, -3, -2) to make sure `m_un[0]`, `m_un[1]`, and `m_un[2]` are written in the correct order.
> 
> There is one last set of failing tests that needed to be fixed:
> 
> ```shell
> self = <astropy.cosmology.flrw.tests.test_wpwazpcdm.TestwpwaCDM object at 0x12b6f0ac0>, cosmo_cls = <class 'astropy.cosmology.flrw.wpwazpcdm.wpwaCDM'>
> cosmo = wpwaCDM(name="ABCMeta", H0=70.0 km / (Mpc s), Om0=0.27, Ode0=0.73, wp=-0.9, wa=0.2, zp=0.5 redshift, Tcmb0=3.0 K, Neff=3.04, m_nu=[0. 0. 0.] eV, Ob0=0.03)
> read = <astropy.cosmology.connect.CosmologyRead object at 0x1693be8c0>, write = <astropy.cosmology.connect.CosmologyWrite object at 0x1693bfa00>
> tmp_path = PosixPath('/private/var/folders/t3/4lbhlq652z16wh1ldqsfr04m0000gn/T/pytest-of-mubinmanasia/pytest-28/test_readwrite_mrt_instance9')
> 
>     def test_readwrite_mrt_instance(
>         self, cosmo_cls, cosmo, read, write, tmp_path
>     ):
>         """Test cosmology -> ascii.mrt -> cosmology."""
>         fp = tmp_path / "test_readwrite_mrt_instance.mrt"
>     
>         # ------------
>         # To Table
>     
>         write(fp, format="ascii.mrt")
>     
>         # some checks on the saved file
>         tbl = Table.read(fp, format="ascii.mrt")
>         assert tbl["name"] == cosmo.name
>     
>         # ------------
>         # From Table
>     
>         tbl["mismatching"] = "will error"
>         tbl.write(fp, format="ascii.mrt", overwrite=True)
>     
>         # tests are different if the last argument is a **kwarg
>         if tuple(cosmo._init_signature.parameters.values())[-1].kind == 4:
>             got = read(fp, format="ascii.mrt")
>     
>             assert got.__class__ is cosmo_cls
>             assert got.name == cosmo.name
>             assert "mismatching" not in got.meta
>     
>             return  # don't continue testing
>     
>         # read with mismatching parameters errors
>         with pytest.raises(TypeError, match="there are unused parameters"):
>             read(fp, format="ascii.mrt")
>     
>         # unless mismatched are moved to meta
> >       got = read(fp, format="ascii.mrt", move_to_meta=True)
> 
> cosmology/io/tests/test_mrt.py:89: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> cosmology/connect.py:86: in __call__
>     cosmo = self.registry.read(self._cls, *args, **kwargs)
> io/registry/core.py:218: in read
>     data = reader(*args, **kwargs)
> cosmology/io/mrt.py:77: in read_mrt
>     return from_table(
> cosmology/io/table.py:138: in from_table
>     return from_row(row, move_to_meta=move_to_meta, cosmology=cosmology)
> cosmology/io/row.py:80: in from_row
>     return from_mapping(mapping, move_to_meta=move_to_meta, cosmology=cosmology)
> cosmology/io/mapping.py:114: in from_mapping
>     return cosmology(*ba.args, **ba.kwargs)
> cosmology/flrw/wpwazpcdm.py:141: in __init__
>     self.zp = zp
> cosmology/parameter.py:138: in __set__
>     value = self.validate(cosmology, copy.deepcopy(value))
> cosmology/parameter.py:186: in validate
>     return self.fvalidate(cosmology, self, value)
> cosmology/parameter.py:348: in _validate_with_unit
>     value = u.Quantity(value, param.unit)
> units/quantity.py:445: in __new__
>     value = value.to(unit)
> units/quantity.py:935: in to
>     value = self._to_value(unit, equivalencies)
> units/quantity.py:888: in _to_value
>     return self.unit.to(
> units/core.py:1195: in to
>     return self._get_converter(Unit(other), equivalencies)(value)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> self = UnrecognizedUnit(redshift), other = Unit("redshift"), equivalencies = []
> 
>     def _get_converter(self, other, equivalencies=None):
>         self._normalize_equivalencies(equivalencies)
> >       raise ValueError(
>             f"The unit {self.name!r} is unrecognized.  It can not be converted "
>             "to other units."
>         )
> E       ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
> 
> units/core.py:2015: ValueError
> ------------------------------------------------------------------------------------------- Captured stdout call -------------------------------------------------------------------------------------------
> cosmology   name       H0      Om0  Ode0 ... Neff m_nu[0] m_nu[1] m_nu[2] Ob0 
>                   km / (Mpc s)           ...         eV      eV      eV       
> --------- ------- ------------ ---- ---- ... ---- ------- ------- ------- ----
>   wpwaCDM ABCMeta         70.0 0.27 0.73 ... 3.04     0.0     0.0     0.0 0.03
> ```
> 
> ```shell
> FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestwpwaCDM::test_readwrite_mrt_instance - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
> FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestwpwaCDM::test_readwrite_mrt_subclass_partial_info - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
> FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestwpwaCDM::test_readwrite_complete_info[ascii.mrt-False-True] - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
> FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestwpwaCDM::test_readwrite_from_subclass_complete_info[ascii.mrt-False-True] - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
> FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestFlatwpwaCDM::test_readwrite_mrt_instance - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
> FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestFlatwpwaCDM::test_readwrite_mrt_subclass_partial_info - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
> FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestFlatwpwaCDM::test_readwrite_complete_info[ascii.mrt-False-True] - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
> FAILED cosmology/flrw/tests/test_wpwazpcdm.py::TestFlatwpwaCDM::test_readwrite_from_subclass_complete_info[ascii.mrt-False-True] - ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.
> ```

@nstarman Do have an idea on how can I tackle this ```ValueError: The unit 'redshift' is unrecognized.  It can not be converted to other units.``` It come from this line ```from_table(table, index=index, move_to_meta=move_to_meta, cosmology=cosmology)```.  I tried using```u.add_enabled_units(cu)```  but it doesn’t help.

**Comment by nstarman** [comment]
Ok. I found a workaround while diagnosing the issue!

**Pull request closed without merging**